.github/workflows/build_and_run_example_test.yml

```
@@ -0,0 +1,90 @@
+name: Build Websockets
+
+on: [push, pull_request]
+
+jobs:
+  build:
+    strategy:
+      matrix:
+        idf_ver: ["latest"]
+        idf_target: ["esp32"]
+
+    runs-on: ubuntu-20.04
+    container: espressif/idf:${{ matrix.idf_ver }}
+    steps:
+      - name: Checkout esp-protocols
+        uses: actions/checkout@master
+        with:
+          path: esp-protocols
+      - name: Build ${{ matrix.example }} with IDF-${{ matrix.idf_ver }} for ${{ matrix.idf_target }}
+        env:
+          IDF_TARGET: ${{ matrix.idf_target }}
+        shell: bash
+        run: |
+          . ${IDF_PATH}/export.sh
+          cd $GITHUB_WORKSPACE/esp-protocols/components/esp_websocket_client/examples/
+          idf.py build
+      - name: Merge binaries
+        env:
+          IDF_TARGET: ${{ matrix.idf_target }}
+        shell: bash
+        run: |
+          . ${IDF_PATH}/export.sh
+          cd $GITHUB_WORKSPACE/esp-protocols/components/esp_websocket_client/examples/build
+          esptool.py --chip ${{ matrix.idf_target }} merge_bin --fill-flash-size 4MB -o flash_image.bin @flash_args
+      - uses: actions/upload-artifact@v2
+        with:
+          name: examples_app_bin_${{ matrix.idf_target }}_${{ matrix.idf_ver }}
+          path: esp-protocols/components/esp_websocket_client/examples/build/*
+
+  run-target:
+    name: Run Example Test on target
+    needs: build
+    strategy:
+      fail-fast: false
+      matrix:
+        idf_ver: ["latest"]
+        idf_target: ["esp32"]
+    runs-on:
+      - ESP32-ETHERNET-KIT
+    container:
+      image: python:3.7-buster
+      options: --privileged # Privileged mode has access to serial ports
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/download-artifact@v2
+        with:
+          name: examples_app_bin_${{ matrix.idf_target }}_${{ matrix.idf_ver }}
+          path: examples/build
+      - name: Install Python packages
+        env:
+          PIP_EXTRA_INDEX_URL: "https://www.piwheels.org/simple"
+        run: | 
+          pip install pytest-embedded-serial-esp~=0.6 pytest-embedded-idf~=0.6
+          pip install junit_xml SimpleWebSocketServer
+      #- name: Download Example Test to target
+      #  run: python -m esptool --chip ${{ matrix.idf_target }} write_flash 0x0 examples/build/flash_image.bin
+      - name: Run Example Test on target
+        run: |
+          cd $GITHUB_WORKSPACE/components/esp_websocket_client/examples
+          pytest -s --junit-xml=./test_app_results_${{ matrix.idf_target }}_${{ matrix.idf_ver }}.xml --embedded-services esp --target=${{ matrix.idf_target }}
```

`-s --embedded-services esp,idf` is used in the pytest.ini. here you're overwritten the cli args.
```suggestion
          pytest --junit-xml=./test_app_results_${{ matrix.idf_target }}_${{ matrix.idf_ver }}.xml --target=${{ matrix.idf_target }}
```

---

you can try run this locally at first. if you don't enable service `idf`, the `app` fixture would be `App` class instance, not `IdfApp`

---

I tried locally 
`pytest --junit-xml=./test_app_results_${{ matrix.idf_target }}_${{ matrix.idf_ver }}.xml --target=${{ matrix.idf_target }}`
And it passed, but in CI I got this error
```
==================================== ERRORS ====================================
______________ ERROR at setup of test_examples_protocol_websocket ______________
/usr/local/lib/python3.7/site-packages/pytest_embedded/plugin.py:1082: in pytest_fixture_setup
    return self._pytest_fixturedef_exec(fixturedef, request, kwargs)
/usr/local/lib/python3.7/site-packages/pytest_embedded/plugin.py:1045: in _pytest_fixturedef_exec
    result = call_fixture_func(fixturefunc, request, kwargs)
/usr/local/lib/python3.7/site-packages/pytest_embedded/plugin.py:3[28](https://github.com/espressif/esp-protocols/runs/5853118205?check_suite_focus=true#step:6:28): in wrapper
    res = func(*args, **kwargs)
/usr/local/lib/python3.7/site-packages/pytest_embedded/plugin.py:897: in serial
    return cls(**_drop_none_kwargs(kwargs))
/usr/local/lib/python3.7/site-packages/pytest_embedded_idf/serial.py:41: in __init__
    raise ValueError(f'Idf app not parsable. Please check if it\'s valid: {self.app.binary_path}')
E   ValueError: Idf app not parsable. Please check if it's valid: None
```

---

I think it would be helpful if you run pytest with `--log-cli-level DEBUG` on CI. and we could see why the binary path is None.

---

yes. seems that the artifacts not downloaded correctly? https://github.com/espressif/esp-protocols/runs/5870212562?check_suite_focus=true#step:6:38

---

Thanks! I have fixed it. And now getting another error on CI. Seems that esp websocket client can't connect to local websocket server
```
2022-04-12 10:41:47 E (23211) esp-tls: [sock=54] select() timeout
2022-04-12 10:41:47 E (23211) TRANSPORT_BASE: Failed to open a new connection: 32774
2022-04-12 10:41:47 E (23211) TRANSPORT_WS: Error connecting to host 192.168.0.2:4455
2022-04-12 10:41:47 E (23211) WEBSOCKET_CLIENT: Error transport connect
```